docs/design-documents/20220309-opencdc.md

```diff
@@ -0,0 +1,349 @@
++		Key DataWithSchema
++		// Payload contains the data and schema of the payload after the change.
++		Payload DataWithSchema
++	}
+}
+
++ // Operation defines what operation triggered the creation of the record.
++ type Operation string
++
++ const (
++ 	OperationInsert Operation = "insert"
++ 	OperationUpdate Operation = "update"
++ 	OperationDelete Operation = "delete"
++ 	OperationSnapshot Operation = "snapshot"
++ )
+
++ // DataWithSchema holds some data and a schema describing the data.
++ type DataWithSchema interface {
++ 	// Schema returns the schema describing Data or nil if none is set.
++ 	Schema() Schema
++ 	// Data returns raw or structured data or nil if none is set.
++ 	Data() Data
++ }
+```
+
+Notable changes:
+
+- `Operation` is added to denote what operation triggered the creation of this record.
+- `CreatedAt` was moved to metadata (see [standard metadata fields](#standard-metadata-fields)).
+- `Key` and `Payload` have been moved under `Before` and `After` to have the ability to represent an updated record.
+- `DataWithSchema` was introduced to create a way of attaching a schema to some data.
+
+#### Standard metadata fields
+
+We propose to define a list of standard metadata fields, like:
+
+- `opencdc.createdAt` - This was previously part of the record but would be moved to metadata as it is an optional field not critical to the operation of Conduit. It represents the time when the change occurred in the source system. If that's impossible to find out, then it should be empty.
+- `opencdc.readAt` - It represents the time the change was detected by the connector.
+
+Conduit could define its own standard metadata fields that it populates automatically, like:
+
+- `conduit.version` - Version of Conduit that processed this record.
+- `conduit.plugin.name` - Name of the Conduit plugin that produced this record.
+- `conduit.plugin.version` - Version of the Conduit plugin that produced this record.
+
+More standard fields can be added. The SDK would provide an easy way of retrieving standard OpenCDC and Conduit metadata fields.
+
+#### Schema format
+
+It should be possible to attach a schema to a record key/payload for at least these formats:
+
+- JSON (JSONSchema)
+- Protobuf
+- Avro
+
+Additionally we might want to support an OpenCDC schema format that describes structured data (See [scenario 4](#scenario-4-connector-ships-structured-data-with-schema)).
+
+### Option 2
+
+OpenCDC gets defined the same as the Debezium record, we change our record to match OpenCDC.
+
+#### Pros
+
+- We are Debezium compatible out of the box.
+- When people work with our internal Record it makes more sense because it is actually the same as an OpenCDC record.
+- If our record is marshalled into JSON (or avro or proto) it's OpenCDC compatible by default.
+
+#### Cons
+
+- Breaking change for connectors, the record will change.
+- We are tied to the Debezium format (less flexibility).
```

Comment:
This feels like a pretty huge con to me. Again, though, I think this comes down to our philosophy for transformers. Are they meant to be minimal data alteration tools, or are they meant to be very capable data massaging and sculpting tools?

Reply:
Depends on what you mean by "very capable" 😄  I imagine transforms should let you alter the OpenCDC record in any way you want, the only constraint is that the output of the transform is still an OpenCDC record.